Does not pass tests

Hello,

First, thank you for these great tools. I installed nix on my debian system, and I got some troubles when I try to compile the nix sources.

**Step to reproduce**
Install the binary distribution of nix on debian sid:

    $ bash <(curl https://nixos.org/nix/install)

Then load the environment using the command that is prompted at the end, and run the following commands:

    nix-env -i git
    mkdir /tmp/compile_nix_git
    cd /tmp/compile_nix_git
    git clone https://github.com/NixOS/nix
    cd nix
    nix-build release.nix -A build.x86_64-linux

Then, after some times, it fails to build, because it fails to pass two tests: `tests/linux-sandbox.sh` and `test/build-remote.sh`, with the same error:

    error: cloning builder process: Operation not permitted

Do you know how to solve this problem ?

Thank you !

NB: Here is the last part of the output of the script, and you can find [here](https://pastebin.com/nzSFDyQv) most of the output of the script. 

```
running test tests/placeholders.sh... [PASS]
running test tests/nix-shell.sh... [PASS]
running test tests/linux-sandbox.sh... [FAIL]
    + clearStore
    + echo 'clearing store...'
    clearing store...
    + chmod -R +w /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store
    + rm -rf /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store
    + mkdir /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store
    + rm -rf /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/var/nix
    + mkdir /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/var/nix
    + nix-store --init
    + clearProfiles
    + profiles=/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/var/nix/profiles
    + rm -rf /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/var/nix/profiles
    ++ uname
    + [[ Linux != Linux ]]
    + [[ ! /nix/store/h404wfcz8rzzlq8vr4z7plcijwzfci72-bash-4.4-p12/bin/bash =~ /nix/store ]]
    + chmod -R u+w /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store0
    chmod: cannot access '/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store0': No such file or directory
    + true
    + rm -rf /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store0
    + export NIX_STORE_DIR=/my/store
    + NIX_STORE_DIR=/my/store
    + export 'NIX_REMOTE=local?root=/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store0'
    + NIX_REMOTE='local?root=/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store0'
    ++ nix-build dependencies.nix --no-out-link --option sandbox-paths /nix/store
    these derivations will be built:
      /my/store/jg37xazcx2r631lf42qbawgb91dwrg99-dependencies-input-2.drv
      /my/store/pzasi0nync7la5rz3sjddrlflpwfsl8g-dependencies-input-1.drv
      /my/store/892g44imcdgz0ijvd8r9gax3h3jmfcvz-dependencies.drv
    building path(s) '/my/store/1zp5618gfawjdnhv3idhfsg5vwq4274v-dependencies-input-1'
    error: cloning builder process: Operation not permitted
    error: unable to start build process
    + outPath=
running test tests/build-remote.sh... [FAIL]
    + clearStore
    + echo 'clearing store...'
    clearing store...
    + chmod -R +w /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store
    + rm -rf /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store
    + mkdir /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store
    + rm -rf /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/var/nix
    + mkdir /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/var/nix
    + nix-store --init
    + clearProfiles
    + profiles=/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/var/nix/profiles
    + rm -rf /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/var/nix/profiles
    ++ uname
    + [[ Linux != Linux ]]
    + [[ ! /nix/store/h404wfcz8rzzlq8vr4z7plcijwzfci72-bash-4.4-p12/bin/bash =~ /nix/store ]]
    + chmod -R u+w /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store0
    + chmod -R u+w /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store1
    chmod: cannot access '/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store1': No such file or directory
    + true
    + rm -rf /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store0 /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store1
    + export NIX_CONF_DIR=/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/etc2
    + NIX_CONF_DIR=/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/etc2
    + mkdir -p /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/etc2
    + echo '
    sandbox-paths = /nix/store
    sandbox-build-dir = /build-tmp
    '
    ++ nix-build build-hook.nix --no-out-link -j0 --option builders 'local?root=/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store0; local?root=/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store1 - - 1 1 foo'
    these derivations will be built:
      /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/8m7qvd39iy8i81ai041kjbslq7vxpcfv-build-hook-input-2.drv
      /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/lynjvq2mq96163h8vxfqdyxyhr84i41h-build-hook-input-1.drv
      /tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/92kzd7wak16kz8l54jjvvgr99s77y8qg-build-hook.drv
    warning: unknown setting 'log-fd'
    warning: unknown setting 'max-connections'
    building '/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/lynjvq2mq96163h8vxfqdyxyhr84i41h-build-hook-input-1.drv'...
    warning: unknown setting 'log-fd'
    warning: unknown setting 'max-connections'
    building '/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/8m7qvd39iy8i81ai041kjbslq7vxpcfv-build-hook-input-2.drv'...
    copying 1 paths...
    copying path '/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/jzkfb7c4fk8444irrw5xa8cpa8pqjgd9-dependencies.builder1.sh'...
    building '/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/lynjvq2mq96163h8vxfqdyxyhr84i41h-build-hook-input-1.drv' on 'local?root=/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store1'
    building path(s) '/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/crznwi5x1pzzm4ahqh40nf7vripiglbd-build-hook-input-1'
    error: cloning builder process: Operation not permitted
    error: build of '/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/lynjvq2mq96163h8vxfqdyxyhr84i41h-build-hook-input-1.drv' on 'local?root=/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store1' failed: unable to start build process
    copying 1 paths...
    cannot build derivation '/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/92kzd7wak16kz8l54jjvvgr99s77y8qg-build-hook.drv': 1 dependencies couldn't be built
    error: build of '/tmp/nix-build-nix-1.12pre1234_abcdef.drv-0/nix-test/store/92kzd7wak16kz8l54jjvvgr99s77y8qg-build-hook.drv' failed
    + outPath=
running test tests/nar-index.sh... [PASS]
2 out of 42 tests failed 
make: *** [mk/tests.mk:12: installcheck] Error 1
build time elapsed:  0m0.104s 0m0.091s 9m14.999s 0m26.938s
builder for ‘/nix/store/fswqs9k2nl2826w0r9n23xvnhdfnmcdj-nix-1.12pre1234_abcdef.drv’ failed with exit code 2
error: build of ‘/nix/store/fswqs9k2nl2826w0r9n23xvnhdfnmcdj-nix-1.12pre1234_abcdef.drv’ failed
```

--- a/tests/build-remote.sh
+++ b/tests/build-remote.sh
@@ -2,7 +2,7 @@ source common.sh
 
 clearStore
 
-if [[ $(uname) != Linux ]]; then exit; fi
+if ! canUseSandbox; then exit; fi
 if [[ ! $SHELL =~ /nix/store ]]; then exit; fi
 
 chmod -R u+w $TEST_ROOT/store0 || true

--- a/tests/common.sh.in
+++ b/tests/common.sh.in
@@ -87,6 +87,24 @@ killDaemon() {
     trap "" EXIT
 }
 
+canUseSandbox() {
+    if [[ $(uname) != Linux ]]; then return 1; fi
+
+    if [ ! -L /proc/self/ns/user ]; then
+        echo "Kernel doesn't support user namespaces, skipping this test..."
+        return 1
+    fi
+
+    if [ -e /proc/sys/kernel/unprivileged_userns_clone ]; then
+        if [ "$(cat /proc/sys/kernel/unprivileged_userns_clone)" != 1 ]; then
+            echo "Unprivileged user namespaces disabled by sysctl, skipping this test..."
+            return 1
+        fi
+    fi
+
+    return 0
+}
+
 fail() {
     echo "$1"
     exit 1

--- a/tests/linux-sandbox.sh
+++ b/tests/linux-sandbox.sh
@@ -2,7 +2,7 @@ source common.sh
 
 clearStore
 
-if [[ $(uname) != Linux ]]; then exit; fi
+if ! canUseSandbox; then exit; fi
 
 # Note: we need to bind-mount $SHELL into the chroot. Currently we
 # only support the case where $SHELL is in the Nix store, because

--- a/tests/run.sh
+++ b/tests/run.sh
@@ -6,24 +6,23 @@ clearCache
 nix run -f run.nix hello -c hello | grep 'Hello World'
 nix run -f run.nix hello -c hello NixOS | grep 'Hello NixOS'
 
-if [[ $(uname) = Linux ]]; then
+if ! canUseSandbox; then exit; fi
 
-    chmod -R u+w $TEST_ROOT/store0 || true
-    rm -rf $TEST_ROOT/store0
+chmod -R u+w $TEST_ROOT/store0 || true
+rm -rf $TEST_ROOT/store0
 
-    clearStore
+clearStore
 
-    path=$(nix eval --raw -f run.nix hello)
+path=$(nix eval --raw -f run.nix hello)
 
-    # Note: we need the sandbox paths to ensure that the shell is
-    # visible in the sandbox.
-    nix run --sandbox-build-dir /build-tmp \
-        --sandbox-paths '/nix? /bin? /lib? /lib64? /usr?' \
-        --store $TEST_ROOT/store0 -f run.nix hello -c hello | grep 'Hello World'
+# Note: we need the sandbox paths to ensure that the shell is
+# visible in the sandbox.
+nix run --sandbox-build-dir /build-tmp \
+    --sandbox-paths '/nix? /bin? /lib? /lib64? /usr?' \
+    --store $TEST_ROOT/store0 -f run.nix hello -c hello | grep 'Hello World'
 
-    path2=$(nix run --sandbox-paths '/nix? /bin? /lib? /lib64? /usr?' --store $TEST_ROOT/store0 -f run.nix hello -c $SHELL -c 'type -p hello')
+path2=$(nix run --sandbox-paths '/nix? /bin? /lib? /lib64? /usr?' --store $TEST_ROOT/store0 -f run.nix hello -c $SHELL -c 'type -p hello')
 
-    [[ $path/bin/hello = $path2 ]]
+[[ $path/bin/hello = $path2 ]]
 
-    [[ -e $TEST_ROOT/store0/nix/store/$(basename $path)/bin/hello ]]
-fi
+[[ -e $TEST_ROOT/store0/nix/store/$(basename $path)/bin/hello ]]

